crates/counter-agg/src/lib.rs

```diff
@@ -204,42 +203,56 @@ impl MetricSummary {
     }
 
     pub fn bounds_valid(&self) -> bool {
-        match self.bounds{
-            None => true,  // unbounded contains everything
-            Some(b) => b.contains(self.last.ts) && b.contains(self.first.ts)
-        }
+        self.bounds.contains(self.last.ts) && self.bounds.contains(self.first.ts)
     }
 
-    fn bounds_extend(&mut self, in_bounds:Option<range::I64Range>){
-        match (self.bounds, in_bounds) {
-            (None, _) => {self.bounds = in_bounds},
-            (_, None) => {},
-            (Some(mut a), Some(b)) => {
-                a.extend(&b);
-                self.bounds = Some(a);
-            }
-        };
+    fn bounds_extend(&mut self, in_bounds: range::I64Range) {
+        // TODO(epg): This preserves existing behavior, which seems odd to me:
+        // match (self.bounds, in_bounds) {
+        // If we're unbounded, narrow to in_bounds (that's not "extend").
+        //     (None, _) => self.bounds = in_bounds,
+        if self.bounds.is_infinite() {
+            self.bounds = in_bounds;
+        }
+        // Else if in_bounds is infinite, ignore it!
+        //     (_, None) => {}
+        else if in_bounds.is_infinite() {
+        }
+        // Else, widen or narrow at both ends as necessary (makes sense but inconsistent!).
+        //     (Some(mut a), Some(b)) => {
+        else {
+            self.bounds.extend(&in_bounds);
+        }
     }
 
     // based on:  https://github.com/timescale/promscale_extension/blob/d51a0958442f66cb78d38b584a10100f0d278298/src/lib.rs#L208,
     // which is based on:     // https://github.com/prometheus/prometheus/blob/e5ffa8c9a08a5ee4185271c8c26051ddc1388b7a/promql/functions.go#L59
     pub fn prometheus_delta(&self) -> Result<Option<f64>, CounterError>{
-        if self.bounds.is_none() || !self.bounds_valid() ||  self.bounds.unwrap().has_infinite() {
+        if !self.bounds_valid() {
             return Err(CounterError::BoundsInvalid);
         }
+        let (left, right);
+        match self.bounds.both() {
+            None => return Err(CounterError::BoundsInvalid),
+            Some((l, r)) => {
+                left = l;
+                right = r;
+            }
+        }
```

Comment:
Seems there really should be a cleaner way to do this `unwrap_or_return` behavior, but if there is I don't know it.